src/jmteb/evaluators/clustering/evaluator.py

```diff
@@ -22,80 +24,95 @@
+        val_dataset: ClusteringDataset,
+        test_dataset: ClusteringDataset,
     ) -> None:
-        self.dataset = dataset
-        self.k = k
-        self.k_means_clustering = k_means_clustering
-        self.hierarchical_clustering = hierarchical_clustering
+        self.val_dataset = val_dataset
+        self.test_dataset = test_dataset
         self.main_metric = "v_measure_score"
 
     def __call__(
         self, model: TextEmbedder, cache_dir: str | PathLike[str] | None = None, overwrite_cache: bool = False
     ) -> EvaluationResults:
         if cache_dir is not None:
             Path(cache_dir).mkdir(parents=True, exist_ok=True)
-        embeddings = model.batch_encode_with_cache(
-            [item.text for item in self.dataset],
-            cache_path=Path(cache_dir) / "embeddings.bin" if cache_dir is not None else None,
+
+        logger.info("Converting validation data to embeddings...")
+        val_embeddings = model.batch_encode_with_cache(
+            [item.text for item in self.val_dataset],
+            cache_path=Path(cache_dir) / "val_embeddings.bin" if cache_dir is not None else None,
             overwrite_cache=overwrite_cache,
         )
-        labels = [item.label for item in self.dataset]
-        n_clusters = len(set(labels))
-
-        results: dict[str, dict[str, float]] = {}
+        val_labels = [item.label for item in self.val_dataset]
 
-        logger.info("Fitting clustering model...")
-        clustering_models = []
-        if self.k_means_clustering:
-            clustering_models.append(
-                MiniBatchKMeans(
-                    n_clusters=n_clusters,
-                    n_init="auto",
-                )
+        logger.info("Converting test data to embeddings...")
+        if self.val_dataset == self.test_dataset:
+            test_embeddings = val_embeddings
+            test_labels = val_labels
+        else:
+            test_embeddings = model.batch_encode_with_cache(
+                [item.text for item in self.test_dataset],
+                cache_path=Path(cache_dir) / "test_embeddings.bin" if cache_dir is not None else None,
+                overwrite_cache=overwrite_cache,
             )
-        if self.hierarchical_clustering:
-            clustering_models.extend(
-                [
-                    AgglomerativeClustering(n_clusters=n_clusters),
-                    BisectingKMeans(n_clusters=n_clusters),
-                    Birch(n_clusters=n_clusters),
-                ]
+            test_labels = [item.label for item in self.test_dataset]
+
+        n_clusters = len(set(test_labels))
+        clustering_models = {
+            type(model).__name__: model
+            for model in (
+                MiniBatchKMeans(n_clusters=n_clusters, n_init="auto"),
+                AgglomerativeClustering(n_clusters=n_clusters),
+                BisectingKMeans(n_clusters=n_clusters),
+                Birch(n_clusters=n_clusters),
             )
```

Comment:
deepcopy すると、後の処理がコピー前のオブジェクトの状態に依存する？ので、少し不安ですね。
よりストレートに、ここは constructor にするなど。

```suggestion
            for model_constructor in (
                lambda: MiniBatchKMeans(n_clusters=n_clusters, n_init="auto"),
                lambda: AgglomerativeClustering(n_clusters=n_clusters),
                lambda: BisectingKMeans(n_clusters=n_clusters),
                lambda: Birch(n_clusters=n_clusters),
            )
```

これで後に `model = model_constructor()` とすると deepcopy しなくて済むはず…